pkg/await/awaiters.go

```diff
@@ -16,15 +16,149 @@ package await
+	extensionsV1Beta1Deployment                 = "extensions/v1beta1/Deployment"
+	extensionsV1Beta1Ingress                    = "extensions/v1beta1/Ingress"
+	rbacAuthorizationV1ClusterRole              = "rbac.authorization.k8s.io/v1/ClusterRole"
+	rbacAuthorizationV1ClusterRoleBinding       = "rbac.authorization.k8s.io/v1/ClusterRoleBinding"
+	rbacAuthorizationV1Role                     = "rbac.authorization.k8s.io/v1/Role"
+	rbacAuthorizationV1RoleBinding              = "rbac.authorization.k8s.io/v1/RoleBinding"
+	rbacAuthorizationV1Alpha1ClusterRole        = "rbac.authorization.k8s.io/v1alpha1/ClusterRole"
+	rbacAuthorizationV1Alpha1ClusterRoleBinding = "rbac.authorization.k8s.io/v1alpha1/ClusterRoleBinding"
+	rbacAuthorizationV1Alpha1Role               = "rbac.authorization.k8s.io/v1alpha1/Role"
+	rbacAuthorizationV1Alpha1RoleBinding        = "rbac.authorization.k8s.io/v1alpha1/RoleBinding"
+	rbacAuthorizationV1Beta1ClusterRole         = "rbac.authorization.k8s.io/v1beta1/ClusterRole"
+	rbacAuthorizationV1Beta1ClusterRoleBinding  = "rbac.authorization.k8s.io/v1beta1/ClusterRoleBinding"
+	rbacAuthorizationV1Beta1Role                = "rbac.authorization.k8s.io/v1beta1/Role"
+	rbacAuthorizationV1Beta1RoleBinding         = "rbac.authorization.k8s.io/v1beta1/RoleBinding"
+	storageV1StorageClass                       = "storage.k8s.io/v1/StorageClass"
+)
+
+type awaitSpec struct {
+	awaitCreation initAwaiter
+	awaitUpdate   updateAwaiter
+	awaitDeletion deletionAwaiter
+}
+
+var deploymentAwaiter = awaitSpec{
+	awaitCreation: untilAppsDeploymentInitialized,
+	awaitUpdate:   untilAppsDeploymentUpdated,
+	awaitDeletion: untilAppsDeploymentDeleted,
+}
+
+// NOTE: Some GVKs below are blank so that we can distinguish between resource types that we know
+// about, but don't require await logic, vs. resource types that we don't know about.
+
+var awaiters = map[string]awaitSpec{
+	appsV1Deployment:                     deploymentAwaiter,
+	appsV1Beta1Deployment:                deploymentAwaiter,
+	appsV1Beta2Deployment:                deploymentAwaiter,
+	autoscalingV1HorizontalPodAutoscaler: { /* NONE */ },
+	coreV1ConfigMap:                      { /* NONE */ },
+	coreV1LimitRange:                     { /* NONE */ },
+	coreV1Namespace: {
+		awaitDeletion: untilCoreV1NamespaceDeleted,
+	},
+	coreV1PersistentVolume: {
+		awaitCreation: untilCoreV1PersistentVolumeInitialized,
+	},
+	coreV1PersistentVolumeClaim: {
+		awaitCreation: untilCoreV1PersistentVolumeClaimBound,
+	},
+	coreV1Pod: {
+		awaitCreation: untilCoreV1PodInitialized,
+		awaitDeletion: untilCoreV1PodDeleted,
+	},
+	coreV1ReplicationController: {
+		awaitCreation: untilCoreV1ReplicationControllerInitialized,
+		awaitUpdate:   untilCoreV1ReplicationControllerUpdated,
+		awaitDeletion: untilCoreV1ReplicationControllerDeleted,
+	},
+	coreV1ResourceQuota: {
+		awaitCreation: untilCoreV1ResourceQuotaInitialized,
+		awaitUpdate:   untilCoreV1ResourceQuotaUpdated,
+	},
+	coreV1Secret: {},
```

Comment:
Probably worth being consistent on whether to include `/* NONE */`.

Reply:
Potentially worth just definining a `noAwaiters` or similar empty struct once, and then using that in each place where no awaiters are needed?

Reply:
Ah, interesting idea. I'd actually prefer to pass it in this way though. I think package-level variables are an anti-pattern.